// js-simple-utils/cjs/RegExp/password.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

/**
 * @module sRegExp.password
 * @since 0.0.2
 * @description sRegExp模块, 提供正则表达式
 */
/**
 * 密码
 * @static
 * @since 0.0.2
 * @description 基础密码，6位数字
 * @constant {RegExp} PASSWORD
 * @example
 * sRegExp.PASSWORD.test('123456')
 */
const PASSWORD = /^\d{6,}$/;
/**
 * 密码
 * @static
 * @since 0.0.2
 * @description 基础密码，4位数字
 * @constant {RegExp} PASSWORD_NUM_4
 * @example
 * sRegExp.PASSWORD_NUM_4.test('1234')
 */
const PASSWORD_NUM_4 = /^\d{4,}$/;
/**
 * 密码
 * @static
 * @since 0.0.2
 * @description 基础密码，6位数字
 * @constant {RegExp} PASSWORD_NUM_6
 * @example
 * sRegExp.PASSWORD_NUM_6.test('123456')
 */
const PASSWORD_NUM_6 = /^\d{6,}$/;
/**
 * 密码
 * @static
 * @since 0.0.2
 * @description 复杂密码，6～16位
 * @constant {RegExp} PASSWORD_ALL_6_16
 * @example
 * sRegExp.PASSWORD_ALL_6_16.test('123456')
 */
const PASSWORD_ALL_6_16 = /^[\x21-\x7E]{6,16}$/;
/**
 * 判断是否是密码
 * @static
 * @since 0.0.2
 * @param {string} password 密码
 * @returns {boolean} true是/false否
 * @example
 * sRegExp.isPassword('123456')
 */
const isPassword = (password) => PASSWORD.test(password);
/**
 * 判断是否是密码
 * @static
 * @since 0.0.2
 * @description 基础密码，4位数字
 * @param {string} password 密码
 * @returns {boolean} true是/false否
 * @example
 * sRegExp.isPasswordNum4('1234')
 */
const isPasswordNum4 = (password) => PASSWORD_NUM_4.test(password);
/**
 * 判断是否是密码
 * @static
 * @since 0.0.2
 * @description 基础密码，6位数字
 * @param {string} password 密码
 * @returns {boolean} true是/false否
 * @example
 * sRegExp.isPasswordNum6('123456')
 */
const isPasswordNum6 = (password) => PASSWORD_NUM_6.test(password);
/**
 * 判断是否是密码
 * @static
 * @since 0.0.2
 * @description 复杂密码，6～16位
 * @param {string} password 密码
 * @returns {boolean} true是/false否
 * @example
 * sRegExp.isPasswordAll6To16('123456')
 */
const isPasswordAll6To16 = (password) => PASSWORD_ALL_6_16.test(password);
var password = {
    PASSWORD,
    PASSWORD_NUM_4,
    PASSWORD_NUM_6,
    PASSWORD_ALL_6_16,
    isPassword,
    isPasswordNum4,
    isPasswordNum6,
    isPasswordAll6To16,
};

exports.PASSWORD = PASSWORD;
exports.PASSWORD_ALL_6_16 = PASSWORD_ALL_6_16;
exports.PASSWORD_NUM_4 = PASSWORD_NUM_4;
exports.PASSWORD_NUM_6 = PASSWORD_NUM_6;
exports["default"] = password;
exports.isPassword = isPassword;
exports.isPasswordAll6To16 = isPasswordAll6To16;
exports.isPasswordNum4 = isPasswordNum4;
exports.isPasswordNum6 = isPasswordNum6;
