// js-simple-utils/cjs/Common/index.js
'use strict';

var Store_index = require('../Store/index.js');
var Array_index = require('../Array/index.js');
var Object_index = require('../Object/index.js');
var String_index = require('../String/index.js');
var Number_index = require('../Number/index.js');
var Dom_index = require('../Dom/index.js');
var Function_index = require('../Function/index.js');
var Url_index = require('../Url/index.js');
require('../RegExp/index.js');
require('../RegExp/password.js');
require('../RegExp/phone.js');

/**
 * @since 0.0.1
 * @description 包使用默认引入，起名规则为s${大驼峰子模块名}
 */
/**
 * @module sCommon
 * @since 0.0.1
 * @description sCommon模块, 整合高频通用工具函数
 * @see {@link sStore}
 * @see {@link sArray.isArray}
 * @see {@link sArray.initArray}
 * @see {@link sArray.fill}
 * @see {@link sObject.isTrueValue}
 * @example
 * // first import sCommon
 * import { sCommon } from 'js-simple-utils'
 * // then use sCommon
 * sCommon.isArray('123@gmail.com') // false
 * sCommon.isTrueValue('') // false
 */
/**
 *@ignore
 */
const sCommon = Object.assign(Object.assign({}, Store_index), { 
    // from sArray
    isArray: Array_index.isArray,
    initArray: Array_index.initArray,
    fill: Array_index.fill,
    // from sObject
    isTrueValue: Object_index.isTrueValue,
    // from sString
    padStart: String_index.padStart,
    padEnd: String_index.padEnd,
    // from sNumber
    pad0: Number_index.pad0,
    // from sDom
    stopPropagation: Dom_index.stopPropagation,
    preventDefault: Dom_index.preventDefault,
    stopAndPrevent: Dom_index.stopAndPrevent,
    // from sFunction
    noop: Function_index.noop,
    deepClone: Function_index.deepClone,
    // from sUrl
    getUrlParams: Url_index.getUrlParams,
    setUrlParams: Url_index.setUrlParams });

module.exports = sCommon;
